apis/oam/v1alpha2/types.go

```diff
@@ -0,0 +1,857 @@
+	// reference this component? Doing so would allow us to queue a reconcile
+	// for consuming ApplicationConfigurations when this Component changed.
+}
+
+// +kubebuilder:object:root=true
+
+// A Component describes how an OAM workload kind may be instantiated.
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:JSONPath=".spec.workload.kind",name=WORKLOAD-KIND,type=string
+type Component struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ComponentSpec   `json:"spec,omitempty"`
+	Status ComponentStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// ComponentList contains a list of Component.
+type ComponentList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []Component `json:"items"`
+}
+
+// A ComponentParameterValue specifies a value for a named parameter. The
+// associated component must publish a parameter with this name.
+type ComponentParameterValue struct {
+	// Name of the component parameter to set.
+	Name string `json:"name"`
+
+	// Value to set.
+	Value intstr.IntOrString `json:"value"`
+}
+
+// A ComponentTrait specifies a trait that should be applied to a component.
+type ComponentTrait struct {
+	// A Trait that will be created for the component
+	// +kubebuilder:validation:EmbeddedResource
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Trait runtime.RawExtension `json:"trait"`
+}
+
+// A ScopeReference refers to an OAM scope resource.
+type ScopeReference struct {
+	// APIVersion of the referenced scope.
+	APIVersion string `json:"apiVersion"`
+
+	// Kind of the referenced scope.
+	Kind string `json:"kind"`
+
+	// Name of the referenced scope.
+	Name string `json:"name"`
+
+	// UID of the referenced scope.
+	// +optional
+	UID *types.UID `json:"uid,omitempty"`
+}
+
+// A ComponentScope specifies a scope in which a component should exist.
+type ComponentScope struct {
+	// A ScopeReference must refer to an OAM scope resource.
+	ScopeReference ScopeReference `json:"scopeRef"`
+}
+
+// An ApplicationConfigurationComponent specifies a component of an
+// ApplicationConfiguration. Each component is used to instantiate a workload.
+type ApplicationConfigurationComponent struct {
+	// ComponentName specifies a component of which an ApplicationConfiguration
+	// should consist. The named component must exist.
+	ComponentName string `json:"componentName"`
+
+	// ParameterValues specify values for the the specified component's
+	// parameters. Any parameter required by the component must be specified.
+	// +optional
+	ParameterValues []ComponentParameterValue `json:"parameterValues,omitempty"`
+
+	// Traits of the specified component.
+	// +optional
+	Traits []ComponentTrait `json:"traits,omitempty"`
+
+	// Scopes in which the specified component should exist.
+	// +optional
+	Scopes []ComponentScope `json:"scopes,omitempty"`
```

Comment:
Is it necessary to add the ComponentScope structure? Can we just use [] ScopeReference?

Reply:
It's not strictly necessary, but I've nested it for future extensibility. The nesting allows us to add other keys at the same level as the `scopeRef` - this is the same principle as @hongchaodeng suggesting each trait CR be enveloped in a `trait` key, rather than simply an array of trait CRs.